# Week 4

### Table of Contents

1. [pd.Series object](#bullet1)
2. [pd.DataFrame object](#bullet2)
3. [Slicing](#bullet3)
4. [Filtering](#bullet4)
5. [Change column/row order](#bullet5)
6. [Rename & sort](#bullet6)
7. [Adding a column](#bullet7)
8. [.describe() method](#bullet8)
9. [pd.unique function](#bullet9)
10. [Frequency and Crosstab Tables](#bullet10)
11. [Dimensions of a df](#bullet11)
12. [Broadcasting](#bullet12)
13. [Hierarchical Indexing](#bullet13)
14. [groupby function](#bullet14)
15. [Importing and Merging data](#bullet15)
16. [Long and Wide Data](#bullet16)

In [1]:
## importing necessary libraries
import numpy as np
import pandas as pd 

## 1. pd.Series object <a class="anchor" id="bullet1"></a>

A Series is a 1-d list-like object containing a sequence of values and an associated array of data labels, the index. Like lists, they can hold a combination of data types

In [6]:
s1 = pd.Series([1, '2', 3, 4.9, 5]) 
# think of a series as a column of a table
# can contain any data type
s1

0      1
1      2
2      3
3    4.9
4      5
dtype: object

In [7]:
print(type(s1[0]))
print(type(s1[1]))
print(type(s1[2]))
print(type(s1[3]))

<class 'int'>
<class 'str'>
<class 'int'>
<class 'float'>


In [8]:
# can set index, which are row names
s2 = pd.Series([4, 7, -5, 3], index=["d", "b", "a", "c"])
s2

d    4
b    7
a   -5
c    3
dtype: int64

## 2. pd.DataFrame object <a class="anchor" id="bullet2"></a>

In [9]:
# most common ways to intialize a dataframe:

In [10]:
# empty dataframe
df1 = pd.DataFrame()
print(df1)

Empty DataFrame
Columns: []
Index: []


In [11]:
df2 = pd.DataFrame(s1)
df2

0
0    1
1    2
2    3
3  4.9
4    5

In [12]:
# give columns names
df3 = pd.DataFrame(s1, columns=['Value'])
df3

Value
0     1
1     2
2     3
3   4.9
4     5

In [13]:
# from a list of lists
data = [['tom', 10], ['nick', 15], ['juli', 14]]
df4 = pd.DataFrame(data, columns=['Name', 'Age'])
df4

Name  Age
0   tom   10
1  nick   15
2  juli   14

In [14]:
# from a dictionary
data = {'Name': ['tom', 'nick', 'juli'],
       'Age':[10, 15, 14]}
df5 = pd.DataFrame(data)
df5

Name  Age
0   tom   10
1  nick   15
2  juli   14

In [18]:
# can label index/rows, note that indeces do not have to be mutually exclusive
df6 = pd.DataFrame(data, index=['1',
                               '2',
                               '2'])
df6

Name  Age
1   tom   10
2  nick   15
2  juli   14

Check out this link for a complete summary on ways to intialize a dataframe:

https://www.geeksforgeeks.org/different-ways-to-create-pandas-dataframe/

The dataset we will be working with as an example for the rest of the Jupyter Notebook is a built-in dataset imported from seaborn

In [19]:
import seaborn as sns
iris = sns.load_dataset("iris")
iris.head() # lists the first 5 rows without any parameters

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [20]:
iris.tail(10) # in this case, lists the last 10 rows (bc was passed the parameter, 10)

sepal_length  sepal_width  petal_length  petal_width    species
140           6.7          3.1           5.6          2.4  virginica
141           6.9          3.1           5.1          2.3  virginica
142           5.8          2.7           5.1          1.9  virginica
143           6.8          3.2           5.9          2.3  virginica
144           6.7          3.3           5.7          2.5  virginica
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

In [ ]:
 # ?? Add column and index view
    #df.info look at columns and their data types
    
    # Look at data values, which are stored as a 2-D
#df.values

# Look at dimensions of DF
#df.shape


## 3. Slicing <a class="anchor" id="bullet3"></a>

Slicing is grabbing a subset of elements. Indexing is grabbing a specific element.

In [21]:
# standard python ways of slicing 

In [22]:
iris['sepal_length'] # returns the given column as a series, put [[]] to return as a df

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal_length, Length: 150, dtype: float64

In [23]:
iris.sepal_width # use column name as an attribute to return the given column as a series

0      3.5
1      3.0
2      3.2
3      3.1
4      3.6
      ... 
145    3.0
146    2.5
147    3.0
148    3.4
149    3.0
Name: sepal_width, Length: 150, dtype: float64

In [24]:
# can select multiple columns
iris[['sepal_length', 'sepal_width']]

sepal_length  sepal_width
0             5.1          3.5
1             4.9          3.0
2             4.7          3.2
3             4.6          3.1
4             5.0          3.6
..            ...          ...
145           6.7          3.0
146           6.3          2.5
147           6.5          3.0
148           6.2          3.4
149           5.9          3.0

[150 rows x 2 columns]

In [26]:
# selecting rows
iris[14:16] # use the [start:stop] format

sepal_length  sepal_width  petal_length  petal_width species
14           5.8          4.0           1.2          0.2  setosa
15           5.7          4.4           1.5          0.4  setosa

In [27]:
# selecting the first 15 rows
iris[:15]

sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8          4.0           1.2          0.2  setosa

In [19]:
# selecting the last element in the list
iris[-1:]

sepal_length  sepal_width  petal_length  petal_width    species
149           5.9          3.0           5.1          1.8  virginica

In [28]:
# selecting a single element
iris['sepal_length'][0:1]

0    5.1
Name: sepal_length, dtype: float64

In [21]:
# ways to slice using pandas methods:

The two methods are loc and iloc. Loc is label based indexing whereas iloc is integer based indexing.

In [29]:
# iloc[row slicing, column slicing]
iris.iloc[20:27, 4:5] # grabbing 7 rows from the last column

species
20  setosa
21  setosa
22  setosa
23  setosa
24  setosa
25  setosa
26  setosa

In [30]:
iris.loc[[0,10], ['petal_length','species']] # grabs row 0 and 10 with the two column values

petal_length species
0            1.4  setosa
10           1.5  setosa

In [38]:
iris.iloc[0, :] # grabs row 0 for all column values

sepal_length       5.1
sepal_width        3.5
petal_length       1.4
petal_width        0.2
species         setosa
Name: 0, dtype: object

NOTE: Labels must be found in the DataFrame or you will get a KeyError.

In [39]:
# to find a specific data elemt using iloc
# iloc[row, column]
iris.iloc[49, 4]

'setosa'

Refresher: copy vs reference

In [40]:
temp1 = iris # this refers to the iris variable and what it contains
# any changes we do to temp1 will appear in iris

In [41]:
temp2 = iris.copy() # makes a copy of the iris dataframe
# any changes done to temp2 won't affect the original iris variable 

In [42]:
# dropping a row permanently 
temp2.drop(0, inplace=True)

In [43]:
temp2.head()

sepal_length  sepal_width  petal_length  petal_width species
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa

In [30]:
# dropping a column permanently
temp2.drop('species', axis=1, inplace=True) # axis = 1, when wanting to drop column by their names or index labels
temp2.head()

sepal_length  sepal_width  petal_length  petal_width
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2
5           5.4          3.9           1.7          0.4

In [44]:
# can drop multiple 
temp2.drop([5, 10, 15, 145], inplace=True)
temp2.drop(['sepal_length', 'sepal_width'], axis=1, inplace=True)
temp2

petal_length  petal_width    species
1             1.4          0.2     setosa
2             1.3          0.2     setosa
3             1.5          0.2     setosa
4             1.4          0.2     setosa
6             1.4          0.3     setosa
..            ...          ...        ...
144           5.7          2.5  virginica
146           5.0          1.9  virginica
147           5.2          2.0  virginica
148           5.4          2.3  virginica
149           5.1          1.8  virginica

[145 rows x 3 columns]

## 4. Filtering <a class="anchor" id="bullet4"></a>

In [45]:
# conditional filtering

In [46]:
len(iris[iris['species'] == 'setosa']) # condition: species is setosa 

50

In [47]:
# can combine multiple conditions
len(iris[(iris['species'] == 'setosa') & (iris['petal_length'] < 1.5)])

24

In [50]:
# query method

In [51]:
len(iris.query('species == "virginica" & petal_length > 5'))

41

In [52]:
# using loc

In [53]:
len(iris.loc[(iris.species == "virginica")])

50

In [54]:
iris.loc[iris.index[0:5],['petal_length','species']] # selecting the first 5 rows for the given columns

petal_length species
0           1.4  setosa
1           1.4  setosa
2           1.3  setosa
3           1.5  setosa
4           1.4  setosa

In [55]:
# Use the .isin() method to match multiple categories
len(iris[iris.species.isin(['virginica','setosa'])]) # multiple values of a column

100

In [58]:
# negate or not equal to
len(iris[~iris.species.isin(['virginica','setosa'])])

50

A good resource as there are so many ways to filter every df:

https://www.listendata.com/2019/07/how-to-filter-pandas-dataframe.html

## 5. Change column/row order <a class="anchor" id="bullet5"></a>

To change row order, use the .reindex function. But, more commonly you probably want to just sort bassed on column value(s)

In [59]:
iris.sort_values("species")

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
27            5.2          3.5           1.5          0.2     setosa
28            5.2          3.4           1.4          0.2     setosa
29            4.7          3.2           1.6          0.2     setosa
30            4.8          3.1           1.6          0.2     setosa
..            ...          ...           ...          ...        ...
119           6.0          2.2           5.0          1.5  virginica
120           6.9          3.2           5.7          2.3  virginica
121           5.6          2.8           4.9          2.0  virginica
111           6.4          2.7           5.3          1.9  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [74]:
# Change col order - this creates a copy, not a view of the dataset
df = iris[['species', 'sepal_length', 'sepal_width', 'petal_length', 'petal_width']]
df

species  sepal_length  sepal_width  petal_length  petal_width
0       setosa           5.1          3.5           1.4          0.2
1       setosa           4.9          3.0           1.4          0.2
2       setosa           4.7          3.2           1.3          0.2
3       setosa           4.6          3.1           1.5          0.2
4       setosa           5.0          3.6           1.4          0.2
..         ...           ...          ...           ...          ...
145  virginica           6.7          3.0           5.2          2.3
146  virginica           6.3          2.5           5.0          1.9
147  virginica           6.5          3.0           5.2          2.0
148  virginica           6.2          3.4           5.4          2.3
149  virginica           5.9          3.0           5.1          1.8

[150 rows x 5 columns]

In [75]:
df.iloc[0, 1] = 5
df

species  sepal_length  sepal_width  petal_length  petal_width
0       setosa           5.0          3.5           1.4          0.2
1       setosa           4.9          3.0           1.4          0.2
2       setosa           4.7          3.2           1.3          0.2
3       setosa           4.6          3.1           1.5          0.2
4       setosa           5.0          3.6           1.4          0.2
..         ...           ...          ...           ...          ...
145  virginica           6.7          3.0           5.2          2.3
146  virginica           6.3          2.5           5.0          1.9
147  virginica           6.5          3.0           5.2          2.0
148  virginica           6.2          3.4           5.4          2.3
149  virginica           5.9          3.0           5.1          1.8

[150 rows x 5 columns]

In [76]:
iris

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

## 6. Rename & Sort <a class="anchor" id="bullet6"></a>

In [77]:
# change column name permanently 
iris.rename(columns={'species':'Species'}, inplace=True)

In [78]:
# change row names temporarily
iris.rename(index={0:'rank0', 1:'rank1'})

sepal_length  sepal_width  petal_length  petal_width    Species
rank0           5.1          3.5           1.4          0.2     setosa
rank1           4.9          3.0           1.4          0.2     setosa
2               4.7          3.2           1.3          0.2     setosa
3               4.6          3.1           1.5          0.2     setosa
4               5.0          3.6           1.4          0.2     setosa
...             ...          ...           ...          ...        ...
145             6.7          3.0           5.2          2.3  virginica
146             6.3          2.5           5.0          1.9  virginica
147             6.5          3.0           5.2          2.0  virginica
148             6.2          3.4           5.4          2.3  virginica
149             5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [79]:
# sort based on a column
iris.sort_values('sepal_length')

sepal_length  sepal_width  petal_length  petal_width    Species
13            4.3          3.0           1.1          0.1     setosa
42            4.4          3.2           1.3          0.2     setosa
38            4.4          3.0           1.3          0.2     setosa
8             4.4          2.9           1.4          0.2     setosa
41            4.5          2.3           1.3          0.3     setosa
..            ...          ...           ...          ...        ...
122           7.7          2.8           6.7          2.0  virginica
118           7.7          2.6           6.9          2.3  virginica
117           7.7          3.8           6.7          2.2  virginica
135           7.7          3.0           6.1          2.3  virginica
131           7.9          3.8           6.4          2.0  virginica

[150 rows x 5 columns]

## 7. Adding a column <a class="anchor" id="bullet7"></a>

In [83]:
# adding a null column
iris['petal_area'] = np.nan
iris.head()

sepal_length  sepal_width  petal_length  petal_width Species  petal_area
0           5.1          3.5           1.4          0.2  setosa         NaN
1           4.9          3.0           1.4          0.2  setosa         NaN
2           4.7          3.2           1.3          0.2  setosa         NaN
3           4.6          3.1           1.5          0.2  setosa         NaN
4           5.0          3.6           1.4          0.2  setosa         NaN

In [84]:
iris['petal_area'] = iris['petal_length']*iris['petal_width']
iris.head()

sepal_length  sepal_width  petal_length  petal_width Species  petal_area
0           5.1          3.5           1.4          0.2  setosa        0.28
1           4.9          3.0           1.4          0.2  setosa        0.28
2           4.7          3.2           1.3          0.2  setosa        0.26
3           4.6          3.1           1.5          0.2  setosa        0.30
4           5.0          3.6           1.4          0.2  setosa        0.28

## 8. .describe() method <a class="anchor" id="bullet8"></a>

In [86]:
iris

sepal_length  sepal_width  petal_length  petal_width    Species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
..            ...          ...           ...          ...        ...   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
..          ...  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18  

[150 rows x 6 columns]

In [87]:
iris.describe() # calculates some general stats for each column, notice that it drops categorical variables by default 

sepal_length  sepal_width  petal_length  petal_width  petal_area
count    150.000000   150.000000    150.000000   150.000000  150.000000
mean       5.843333     3.057333      3.758000     1.199333    5.794067
std        0.828066     0.435866      1.765298     0.762238    4.712390
min        4.300000     2.000000      1.000000     0.100000    0.110000
25%        5.100000     2.800000      1.600000     0.300000    0.420000
50%        5.800000     3.000000      4.350000     1.300000    5.615000
75%        6.400000     3.300000      5.100000     1.800000    9.690000
max        7.900000     4.400000      6.900000     2.500000   15.870000

In [127]:
iris.describe().loc[['mean', 'std']]

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
mean      5.843333     3.057333      3.758000     1.199333    5.794067   
std       0.828066     0.435866      1.765298     0.762238    4.712390   

      sqrt_petal_area  
mean         2.102905  
std          1.175187

## 9. pd.unique() function <a class="anchor" id="bullet9"></a>

In [89]:
pd.unique(iris.Species) # gives the unique values that appear for the given column

array(['setosa', 'versicolor', 'virginica'], dtype=object)

## 10. Frequency and Crosstab Tables <a class="anchor" id="bullet10"></a>

In [90]:
iris['Species'].value_counts()

setosa        50
versicolor    50
virginica     50
Name: Species, dtype: int64

In [93]:
round(iris['Species'].value_counts(normalize=True),2) # relative frequency table

setosa        0.33
versicolor    0.33
virginica     0.33
Name: Species, dtype: float64

In [94]:
pd.crosstab(index=iris['Species'], columns='count')

col_0       count
Species          
setosa         50
versicolor     50
virginica      50

In [1]:
iris['petal_area_cate'] = np.where(iris['petal_area'] >= np.median(iris['petal_area']), 'Large', 'Small')
pd.crosstab(index=iris['Species'], columns=iris['petal_area_cate'])

NameError: name 'np' is not defined

In [98]:
pd.crosstab(index=iris['Species'], columns=iris['petal_area_cate'], margins=True)

petal_area_cate  Large  Small  All
Species                           
setosa               0     50   50
versicolor          25     25   50
virginica           50      0   50
All                 75     75  150

Helpful resource: https://www.datasciencemadesimple.com/create-frequency-table-of-column-in-pandas-python-2/

## 11. Dimensions of a Dataframe <a class="anchor" id="bullet11"></a>

In [99]:
iris.shape # returns (rows, cols)

(150, 7)

In [100]:
# ways to return the number of rows
print(len(iris.index))
print(iris.shape[0])

150
150


In [101]:
# ways to return the number of columns
print(iris.shape[1])
print(len(iris.columns))

7
7


## 12. Broadcasting <a class="anchor" id="bullet12"></a>

In [102]:
## As in NumPy, pandas supports broadcasting or vectorized operations
iris['sqrt_petal_area'] = np.sqrt(iris['petal_area'])
iris.head()

sepal_length  sepal_width  petal_length  petal_width Species  petal_area  \
0           5.1          3.5           1.4          0.2  setosa        0.28   
1           4.9          3.0           1.4          0.2  setosa        0.28   
2           4.7          3.2           1.3          0.2  setosa        0.26   
3           4.6          3.1           1.5          0.2  setosa        0.30   
4           5.0          3.6           1.4          0.2  setosa        0.28   

  petal_area_cate  sqrt_petal_area  
0           Small         0.529150  
1           Small         0.529150  
2           Small         0.509902  
3           Small         0.547723  
4           Small         0.529150

## 13. Hierachical Indexing <a class="anchor" id="bullet13"></a>

"Hierarchical indexing is an important feature of pandas that enables you to have multiple (two or more) index levels on an axis. Another way of thinking about it is that it provides a way for you to work with higher dimensional data in a lower dimensional form." - _McKinney_

In [103]:
data = pd.Series(np.random.uniform(size=9),
                  index=[["a", "a", "a", "b", "b", "c", "c", "d", "d"],
                         [1, 2, 3, 1, 3, 1, 2, 2, 3]])
data

a  1    0.367510
   2    0.214619
   3    0.340104
b  1    0.882754
   3    0.838943
c  1    0.711609
   2    0.445934
d  2    0.672168
   3    0.173940
dtype: float64

In [104]:
data.index

MultiIndex([('a', 1),
            ('a', 2),
            ('a', 3),
            ('b', 1),
            ('b', 3),
            ('c', 1),
            ('c', 2),
            ('d', 2),
            ('d', 3)],
           )

In [105]:
data['b']

1    0.882754
3    0.838943
dtype: float64

In [106]:
data['b':'c']

b  1    0.882754
   3    0.838943
c  1    0.711609
   2    0.445934
dtype: float64

In [107]:
data.loc[['b', 'd']]

b  1    0.882754
   3    0.838943
d  2    0.672168
   3    0.173940
dtype: float64

In [108]:
data.loc[:,2] # 2 is the name of the second index!

a    0.214619
c    0.445934
d    0.672168
dtype: float64

In [110]:
data.unstack()

1         2         3
a  0.367510  0.214619  0.340104
b  0.882754       NaN  0.838943
c  0.711609  0.445934       NaN
d       NaN  0.672168  0.173940

In [111]:
# columns can also have hierarchical indexing! Refer to mckinney ch 8

## 14. groupby function <a class="anchor" id="bullet14"></a>

"Pandas groupby is used for grouping the data according to the categories and apply a function to the categories. It also helps to aggregate data efficiently." - https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/

In [114]:
iris_grouped = iris.groupby('Species')

In [115]:
iris_grouped.first() # prints the first entry in all the groups formed

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
Species                                                                        
setosa               5.1          3.5           1.4          0.2        0.28   
versicolor           7.0          3.2           4.7          1.4        6.58   
virginica            6.3          3.3           6.0          2.5       15.00   

           petal_area_cate  sqrt_petal_area  
Species                                      
setosa               Small         0.529150  
versicolor           Large         2.565151  
virginica            Large         3.872983

In [116]:
iris_grouped.head() # first 5 entries of every group

sepal_length  sepal_width  petal_length  petal_width     Species  \
0             5.1          3.5           1.4          0.2      setosa   
1             4.9          3.0           1.4          0.2      setosa   
2             4.7          3.2           1.3          0.2      setosa   
3             4.6          3.1           1.5          0.2      setosa   
4             5.0          3.6           1.4          0.2      setosa   
50            7.0          3.2           4.7          1.4  versicolor   
51            6.4          3.2           4.5          1.5  versicolor   
52            6.9          3.1           4.9          1.5  versicolor   
53            5.5          2.3           4.0          1.3  versicolor   
54            6.5          2.8           4.6          1.5  versicolor   
100           6.3          3.3           6.0          2.5   virginica   
101           5.8          2.7           5.1          1.9   virginica   
102           7.1          3.0           5.9          2.1   virginica   
103           6.3          2.9           5.6          1.8   virginica   
104           6.5          3.0           5.8          2.2   virginica   

     petal_area petal_area_cate  sqrt_petal_area  
0          0.28           Small         0.529150  
1          0.28           Small         0.529150  
2          0.26           Small         0.509902  
3          0.30           Small         0.547723  
4          0.28           Small         0.529150  
50         6.58           Large         2.565151  
51         6.75           Large         2.598076  
52         7.35           Large         2.711088  
53         5.20           Small         2.280351  
54         6.90           Large         2.626785  
100       15.00           Large         3.872983  
101        9.69           Large         3.112876  
102       12.39           Large         3.519943  
103       10.08           Large         3.174902  
104       12.76           Large         3.572114

In [69]:
# can group by multiple columns
iris.groupby(['Species', 'petal_area']).value_counts()

Species    petal_area  sepal_length  sepal_width  petal_length  petal_width  sqrt_petal_area
setosa     0.11        4.3           3.0          1.1           0.1          0.331662           1
           0.14        4.8           3.0          1.4           0.1          0.374166           1
                       4.9           3.6          1.4           0.1          0.374166           1
           0.15        4.9           3.1          1.5           0.1          0.387298           1
                       5.2           4.1          1.5           0.1          0.387298           1
                                                                                               ..
virginica  14.25       6.7           3.3          5.7           2.5          3.774917           1
           14.74       7.7           3.8          6.7           2.2          3.839271           1
           15.00       6.3           3.3          6.0           2.5          3.872983           1
           15.25       7.

In [121]:
# can apply aggregate functions
iris.groupby(['Species']).mean()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
Species                                                                        
setosa             5.006        3.428         1.462        0.246      0.3656   
versicolor         5.936        2.770         4.260        1.326      5.7204   
virginica          6.588        2.974         5.552        2.026     11.2962   

            sqrt_petal_area  
Species                      
setosa             0.588753  
versicolor         2.374185  
virginica          3.345777

In [125]:
# Use describe() method but only looking at mean and std
iris.groupby(['Species']).describe().loc[:,(slice(None), ['mean', 'std'])]

sepal_length           sepal_width           petal_length  \
                   mean       std        mean       std         mean   
Species                                                                
setosa            5.006  0.352490       3.428  0.379064        1.462   
versicolor        5.936  0.516171       2.770  0.313798        4.260   
virginica         6.588  0.635880       2.974  0.322497        5.552   

                     petal_width           petal_area            \
                 std        mean       std       mean       std   
Species                                                           
setosa      0.173664       0.246  0.105386     0.3656  0.181155   
versicolor  0.469911       1.326  0.197753     5.7204  1.368403   
virginica   0.551895       2.026  0.274650    11.2962  2.157412   

           sqrt_petal_area            
                      mean       std  
Species                               
setosa            0.588753  0.139128  
versicolor        2.374185  0.292155  
virginica         3.345777  0.322581

## 15. Importing and Merging data <a class="anchor" id="bullet15"></a>

This example was taken from [here](https://realpython.com/pandas-merge-join-and-concat/), where more resources for learning how to merge datasets is available.

In [197]:
# Set path to import the two datasets we'll be merging
path = r'C:\Users\nicka\OneDrive\Desktop\1 Modern Data Structures - GR5072\QMSS-GR5072_Spring2023\Week 4\Data' #r stands for raw, so we can only type one \
path

'C:\\Users\\nicka\\OneDrive\\Desktop\\1 Modern Data Structures - GR5072\\QMSS-GR5072_Spring2023\\Week 4\\Data'

In [198]:
climate_temp = pd.read_csv(path + "\climate_temp.csv")
print(climate_temp.shape)
climate_temp.head()

(127020, 21)


STATION            STATION_NAME ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  DLY-CLDD-BASE72  \
0            -7777  ...                0                0                0   
1            -7777  ...                0                0                0   
2            -7777  ...                0                0                0   
3            -7777  ...                0                0                0   
4            -7777  ...                0                0                0   

   DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  DLY-HTDD-BASE55  \
0            -7777                1                2                6   
1            -7777                1                2                6   
2            -7777                1                2                5   
3            -7777                1                2                5   
4            -7777            -7777                2                5   

   DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0                7               10               15  
1                7               10               15  
2                7               10               15  
3                7               10               15  
4                7               10               15  

[5 rows x 21 columns]

In [195]:
climate_precip = pd.read_csv(path + "\climate_precip.csv")
print(climate_precip.shape)
climate_precip.head()

(151110, 29)


STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101            -6.66   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102            -6.66   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103            -6.66   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104            -6.66   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105            -6.66   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6            -6.66             -666   
1             -666            -66.6            -6.66             -666   
2             -666            -66.6            -6.66             -666   
3            -9999          -9999.0            -6.66            -9999   
4            -9999          -9999.0            -6.66            -9999   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-PRCP-PCTALL-GE100HI  \
0            -66.6            -6.66  ...                        3   
1            -66.6            -6.66  ...                        3   
2            -66.6            -6.66  ...                        3   
3          -9999.0            -6.66  ...                        3   
4          -9999.0            -6.66  ...                        3   

   DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  DLY-SNWD-PCTALL-GE003WI  \
0                    -9999                        0                    -9999   
1                    -9999                        0                    -9999   
2                    -9999                        0                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  \
0                    -9999                    -9999                    -9999   
1                    -9999                    -9999                    -9999   
2                    -9999                    -9999                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  DLY-SNOW-PCTALL-GE050TI  
0                        0                    -9999                    -9999  
1                        0                    -9999                    -9999  
2                        0                    -9999                    -9999  
3                        0                        0                        0  
4                        0                        0                        0  

[5 rows x 29 columns]

### Inner Join

Here, you will use a plain `merge()` call to do an inner join and learn how this can result in a smaller, more focused dataset. First, you will create a new DataFrame object that contains the precipitation data from one station.

In [199]:
precip_one_station = climate_precip.query("STATION == 'GHCND:USC00045721'")
precip_one_station.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
1460  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1461  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
1462  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
1463  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
1464  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

      DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
1460             -666            -66.6             0.16             -666   
1461             -666            -66.6             0.16             -666   
1462             -666            -66.6             0.16             -666   
1463             -666            -66.6             0.16             -666   
1464             -666            -66.6             0.17             -666   

      DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-PRCP-PCTALL-GE100HI  \
1460            -66.6             0.44  ...                       11   
1461            -66.6             0.44  ...                       11   
1462            -66.6             0.45  ...                       11   
1463            -66.6             0.45  ...                       11   
1464            -66.6             0.46  ...                       11   

      DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
1460                        4                        0   
1461                        4                        0   
1462                        4                        0   
1463                        4                        0   
1464                        4                        0   

      DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
1460                        3                        3   
1461                        3                        3   
1462                        3                        3   
1463                        3                        2   
1464                        3                        2   

      DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  \
1460                        9                        6   
1461                       10                        6   
1462                       10                        6   
1463                       10                        6   
1464                       10                        6   

      DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
1460                        0                    -9999   
1461                        0                    -9999   
1462                        0                    -9999   
1463                        0                    -9999   
1464                        0                    -9999   

      DLY-SNOW-PCTALL-GE050TI  
1460                    -9999  
1461                    -9999  
1462                    -9999  
1463                    -9999  
1464                    -9999  

[5 rows x 29 columns]

In [200]:
precip_one_station.shape

(365, 29)

In [201]:
inner_merged = pd.merge(precip_one_station, climate_temp)
inner_merged.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -66.6             0.44  ...                0                0   
1            -66.6             0.44  ...                0                0   
2            -66.6             0.45  ...                0                0   
3            -66.6             0.45  ...                0                0   
4            -66.6             0.46  ...                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0                1                3                6   
1                0                1                3                6   
2                0                1                2                5   
3                0                1                2                5   
4                0                1                2                5   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0               10               12               14               19  
1               10               11               14               19  
2                9               11               14               19  
3                9               11               14               19  
4                9               11               14               19  

[5 rows x 47 columns]

In [202]:
precip_one_station.index

Int64Index([1460, 1461, 1462, 1463, 1464, 1465, 1466, 1467, 1468, 1469,
            ...
            1815, 1816, 1817, 1818, 1819, 1820, 1821, 1822, 1823, 1824],
           dtype='int64', length=365)

In [203]:
climate_temp.index

RangeIndex(start=0, stop=127020, step=1)

In [204]:
inner_merged = pd.merge(precip_one_station, climate_temp) # merge is based on integer index values
inner_merged.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -66.6             0.44  ...                0                0   
1            -66.6             0.44  ...                0                0   
2            -66.6             0.45  ...                0                0   
3            -66.6             0.45  ...                0                0   
4            -66.6             0.46  ...                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0                1                3                6   
1                0                1                3                6   
2                0                1                2                5   
3                0                1                2                5   
4                0                1                2                5   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0               10               12               14               19  
1               10               11               14               19  
2                9               11               14               19  
3                9               11               14               19  
4                9               11               14               19  

[5 rows x 47 columns]

How many rows do you think this merged DataFrame has?

In [205]:
inner_merged.shape

(365, 47)

You get 365 rows because any non-matching rows are discarded in an inner join, which is the default merge method for a `merge()` call, and `precip_one_station` had only 365 rows.

What if you want to merge both full datasets, but specify which columns to join on? In this case, you will use the `on` parameter:

In [206]:
inner_merged_total = pd.merge(
    climate_temp, climate_precip, on=["STATION", "DATE"]
)
inner_merged_total.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  ...  DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  \
0            -7777  ...                        3                    -9999   
1            -7777  ...                        3                    -9999   
2            -7777  ...                        3                    -9999   
3            -7777  ...                        3                        0   
4            -7777  ...                        3                        0   

   DLY-SNWD-PCTALL-GE010WI  DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
0                        0                    -9999                    -9999   
1                        0                    -9999                    -9999   
2                        0                    -9999                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  \
0                    -9999                    -9999                        0   
1                    -9999                    -9999                        0   
2                    -9999                    -9999                        0   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE030TI  DLY-SNOW-PCTALL-GE050TI  
0                    -9999                    -9999  
1                    -9999                    -9999  
2                    -9999                    -9999  
3                        0                        0  
4                        0                        0  

[5 rows x 48 columns]

In [207]:
inner_merged_total.shape

(123005, 48)

In [208]:
inner_merged_total.columns

Index(['STATION', 'STATION_NAME_x', 'ELEVATION', 'LATITUDE', 'LONGITUDE',
       'DATE', 'DLY-CLDD-BASE45', 'DLY-CLDD-BASE50', 'DLY-CLDD-BASE55',
       'DLY-CLDD-BASE57', 'DLY-CLDD-BASE60', 'DLY-CLDD-NORMAL',
       'DLY-CLDD-BASE70', 'DLY-CLDD-BASE72', 'DLY-HTDD-BASE40',
       'DLY-HTDD-BASE45', 'DLY-HTDD-BASE50', 'DLY-HTDD-BASE55',
       'DLY-HTDD-BASE57', 'DLY-HTDD-BASE60', 'DLY-HTDD-NORMAL',
       'STATION_NAME_y', 'DLY-PRCP-25PCTL', 'DLY-SNWD-25PCTL',
       'DLY-SNOW-25PCTL', 'DLY-PRCP-50PCTL', 'DLY-SNWD-50PCTL',
       'DLY-SNOW-50PCTL', 'DLY-PRCP-75PCTL', 'DLY-SNWD-75PCTL',
       'DLY-SNOW-75PCTL', 'MTD-PRCP-NORMAL', 'MTD-SNOW-NORMAL',
       'YTD-PRCP-NORMAL', 'YTD-SNOW-NORMAL', 'DLY-PRCP-PCTALL-GE001HI',
       'DLY-PRCP-PCTALL-GE010HI', 'DLY-PRCP-PCTALL-GE050HI',
       'DLY-PRCP-PCTALL-GE100HI', 'DLY-SNWD-PCTALL-GE001WI',
       'DLY-SNWD-PCTALL-GE010WI', 'DLY-SNWD-PCTALL-GE003WI',
       'DLY-SNWD-PCTALL-GE005WI', 'DLY-SNOW-PCTALL-GE001TI',
       'DLY-SNOW-PCTALL-GE0

You can specify a single _key column_ with a string, or multiple key columns with a list, as in the above example. This results in a DataFrame with 123,005 rows and 48 columns. 

Why 48 columns instead of 47? Because you specified the keys columns to join on, Pandas doesn't try to merge all mergeable columns. This can result in "duplicate" column names, which may or may not have different values. "Duplicate" is in quotes because the columns will actually have new names, by default they are appended with `_x` and `_y`. You can also use the `suffixes` parameter to control what is appended to the column names.

In [209]:
pd.crosstab(index=inner_merged_total['STATION_NAME_x'], columns=inner_merged_total['STATION_NAME_y'])

STATION_NAME_y                    ADIN RANGER STATION CA US  \
STATION_NAME_x                                                
ADIN RANGER STATION CA US                               365   
ALAMEDA NAS CA US                                         0   
ALPINE CA US                                              0   
ALTURAS CA US                                             0   
ALTURAS MUNICIPAL AIRPORT CA US                           0   
...                                                     ...   
WRIGHTWOOD CA US                                          0   
YOSEMITE PARK HEADQUARTERS CA US                          0   
YOSEMITE VILLAGE 12 W CA US                               0   
YOUNTVILLE CA US                                          0   
YREKA CA US                                               0   

STATION_NAME_y                    ALAMEDA NAS CA US  ALPINE CA US  \
STATION_NAME_x                                                      
ADIN RANGER STATION CA US                         0             0   
ALAMEDA NAS CA US                               365             0   
ALPINE CA US                                      0           365   
ALTURAS CA US                                     0             0   
ALTURAS MUNICIPAL AIRPORT CA US                   0             0   
...                                             ...           ...   
WRIGHTWOOD CA US                                  0             0   
YOSEMITE PARK HEADQUARTERS CA US                  0             0   
YOSEMITE VILLAGE 12 W CA US                       0             0   
YOUNTVILLE CA US                                  0             0   
YREKA CA US                                       0             0   

STATION_NAME_y                    ALTURAS CA US  \
STATION_NAME_x                                    
ADIN RANGER STATION CA US                     0   
ALAMEDA NAS CA US                             0   
ALPINE CA US                                  0   
ALTURAS CA US                               365   
ALTURAS MUNICIPAL AIRPORT CA US               0   
...                                         ...   
WRIGHTWOOD CA US                              0   
YOSEMITE PARK HEADQUARTERS CA US              0   
YOSEMITE VILLAGE 12 W CA US                   0   
YOUNTVILLE CA US                              0   
YREKA CA US                                   0   

STATION_NAME_y                    ALTURAS MUNICIPAL AIRPORT CA US  \
STATION_NAME_x                                                      
ADIN RANGER STATION CA US                                       0   
ALAMEDA NAS CA US                                               0   
ALPINE CA US                                                    0   
ALTURAS CA US                                                   0   
ALTURAS MUNICIPAL AIRPORT CA US                               365   
...                                                           ...   
WRIGHTWOOD CA US                                                0   
YOSEMITE PARK HEADQUARTERS CA US                                0   
YOSEMITE VILLAGE 12 W CA US                                     0   
YOUNTVILLE CA US                                                0   
YREKA CA US                                                     0   

STATION_NAME_y                    ANAHEIM CA US  \
STATION_NAME_x                                    
ADIN RANGER STATION CA US                     0   
ALAMEDA NAS CA US                             0   
ALPINE CA US                                  0   
ALTURAS CA US                                 0   
ALTURAS MUNICIPAL AIRPORT CA US               0   
...                                         ...   
WRIGHTWOOD CA US                              0   
YOSEMITE PARK HEADQUARTERS CA US              0   
YOSEMITE VILLAGE 12 W CA US                   0   
YOUNTVILLE CA US                              0   
YREKA CA US                                   0   

STATION_NAME_y                    ANGWIN PACIFIC UNION C

### Outer Join
With the outer join, you will retain rows that don't have matches as well. For this example, you will use the smaller precipitation DataFrame `precip_one_station` with the full `climate_temp` DataFrame and join with `STATION` and `DATE` columns as the key columns. Take a second and think about how many rows you expect the new DataFrame to have.

In [184]:
outer_merged = pd.merge(
    precip_one_station, climate_temp, how="outer", on=["STATION", "DATE"]
)
outer_merged.head()

STATION          STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0           -666.0            -66.6             0.16           -666.0   
1           -666.0            -66.6             0.16           -666.0   
2           -666.0            -66.6             0.16           -666.0   
3           -666.0            -66.6             0.16           -666.0   
4           -666.0            -66.6             0.17           -666.0   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -66.6             0.44  ...                0                0   
1            -66.6             0.44  ...                0                0   
2            -66.6             0.45  ...                0                0   
3            -66.6             0.45  ...                0                0   
4            -66.6             0.46  ...                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0                1                3                6   
1                0                1                3                6   
2                0                1                2                5   
3                0                1                2                5   
4                0                1                2                5   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0               10               12               14               19  
1               10               11               14               19  
2                9               11               14               19  
3                9               11               14               19  
4                9               11               14               19  

[5 rows x 48 columns]

In [185]:
outer_merged.shape

(127020, 48)

If you remember from when you checked the `.shape` attribute of `climate_temp`, you'll see that the number of rows in `outer_merged` matches that. With an outer join, you can expect to have the same number of rows as the larger DataFrame, since none are lost like they are in an inner join. 

### Left Join
Also known as a left outer join. In this join, you will retain rows that don't have matches only on the left (or first) DataFrame to be merged.

In [186]:
left_merged = pd.merge(
    climate_temp, precip_one_station, how="left", on=["STATION", "DATE"]
)
left_merged.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  ...  DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  \
0            -7777  ...                      NaN                      NaN   
1            -7777  ...                      NaN                      NaN   
2            -7777  ...                      NaN                      NaN   
3            -7777  ...                      NaN                      NaN   
4            -7777  ...                      NaN                      NaN   

   DLY-SNWD-PCTALL-GE010WI  DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE030TI  DLY-SNOW-PCTALL-GE050TI  
0                      NaN                      NaN  
1                      NaN                      NaN  
2                      NaN                      NaN  
3                      NaN                      NaN  
4                      NaN                      NaN  

[5 rows x 48 columns]

In [187]:
left_merged.shape

(127020, 48)

Here, you see that the number of rows in the resulting DataFrame matches that of the rows in the `climate_temp` DataFrame. What if we switched the positions of the two DataFrames that we are merging?

In [189]:
left_merged_reversed = pd.merge(
    precip_one_station, climate_temp, how="left", on=["STATION", "DATE"]
)
left_merged_reversed.head()

STATION          STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -66.6             0.44  ...                0                0   
1            -66.6             0.44  ...                0                0   
2            -66.6             0.45  ...                0                0   
3            -66.6             0.45  ...                0                0   
4            -66.6             0.46  ...                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0                1                3                6   
1                0                1                3                6   
2                0                1                2                5   
3                0                1                2                5   
4                0                1                2                5   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0               10               12               14               19  
1               10               11               14               19  
2                9               11               14               19  
3                9               11               14               19  
4                9               11               14               19  

[5 rows x 48 columns]

In [190]:
left_merged_reversed.shape

(365, 48)

### Right Join
This works the same as the left join, however non-matching rows are only retained in the _right_ DataFrame. In the next example, you will recreate the `left_merged` DataFrame but with a right join.

In [191]:
right_merged = pd.merge(
    precip_one_station, climate_temp, how="right", on=["STATION", "DATE"]
)
right_merged.head()

STATION STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099            NaN  20100101              NaN   
1  GHCND:USC00049099            NaN  20100102              NaN   
2  GHCND:USC00049099            NaN  20100103              NaN   
3  GHCND:USC00049099            NaN  20100104              NaN   
4  GHCND:USC00049099            NaN  20100105              NaN   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0              NaN              NaN  ...                0                0   
1              NaN              NaN  ...                0                0   
2              NaN              NaN  ...                0                0   
3              NaN              NaN  ...                0                0   
4              NaN              NaN  ...                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0                6                7               10               15  
1                6                7               10               15  
2                5                7               10               15  
3                5                7               10               15  
4                5                7               10               15  

[5 rows x 48 columns]

In [170]:
right_merged.shape

(127020, 48)

## 16. Long and wide data <a class="anchor" id="bullet16"></a>

Sometimes, you need to use data in a __"wide"__ or __"long"__ format. You can switch back and forth fairly easily in pandas

### Switching to Long or Wide format

In [17]:
path = r'C:\Users\nicka\OneDrive\Desktop\1 Modern Data Structures - GR5072\QMSS-GR5072_Spring2023\Week 4\Data' #r stands for raw, so we can only type one \
path

'C:\\Users\\nicka\\OneDrive\\Desktop\\1 Modern Data Structures - GR5072\\QMSS-GR5072_Spring2023\\Week 4\\Data'

In [18]:
data = pd.read_csv(path + "/macrodata.csv")
data = data.loc[:, ["year", "quarter", "realgdp", "infl", "unemp"]]
data

year  quarter    realgdp  infl  unemp
0    1959        1   2710.349  0.00    5.8
1    1959        2   2778.801  2.34    5.1
2    1959        3   2775.488  2.74    5.3
3    1959        4   2785.204  0.27    5.6
4    1960        1   2847.699  2.31    5.2
..    ...      ...        ...   ...    ...
198  2008        3  13324.600 -3.16    6.0
199  2008        4  13141.920 -8.79    6.9
200  2009        1  12925.410  0.94    8.1
201  2009        2  12901.504  3.37    9.2
202  2009        3  12990.341  3.56    9.6

[203 rows x 5 columns]

In [19]:
periods = pd.PeriodIndex(year=data.pop("year"),
                             quarter=data.pop("quarter"),
                             name="date")

periods
data.index = periods.to_timestamp("D")
data.head()

realgdp  infl  unemp
date                             
1959-01-01  2710.349  0.00    5.8
1959-04-01  2778.801  2.34    5.1
1959-07-01  2775.488  2.74    5.3
1959-10-01  2785.204  0.27    5.6
1960-01-01  2847.699  2.31    5.2

In [20]:
data = data.reindex(columns=["realgdp", "infl", "unemp"])
data.columns.name = "item"
data.head()

item         realgdp  infl  unemp
date                             
1959-01-01  2710.349  0.00    5.8
1959-04-01  2778.801  2.34    5.1
1959-07-01  2775.488  2.74    5.3
1959-10-01  2785.204  0.27    5.6
1960-01-01  2847.699  2.31    5.2

In [21]:
long_data = (data.stack()
              .reset_index()
              .rename(columns={0: "value"}))

In [22]:
long_data[:10]

date     item     value
0 1959-01-01  realgdp  2710.349
1 1959-01-01     infl     0.000
2 1959-01-01    unemp     5.800
3 1959-04-01  realgdp  2778.801
4 1959-04-01     infl     2.340
5 1959-04-01    unemp     5.100
6 1959-07-01  realgdp  2775.488
7 1959-07-01     infl     2.740
8 1959-07-01    unemp     5.300
9 1959-10-01  realgdp  2785.204

In [24]:
pivoted = long_data.pivot(index="date", columns="item", values="value") # Switch back to original / wide format
pivoted

item        infl    realgdp  unemp
date                              
1959-01-01  0.00   2710.349    5.8
1959-04-01  2.34   2778.801    5.1
1959-07-01  2.74   2775.488    5.3
1959-10-01  0.27   2785.204    5.6
1960-01-01  2.31   2847.699    5.2
...          ...        ...    ...
2008-07-01 -3.16  13324.600    6.0
2008-10-01 -8.79  13141.920    6.9
2009-01-01  0.94  12925.410    8.1
2009-04-01  3.37  12901.504    9.2
2009-07-01  3.56  12990.341    9.6

[203 rows x 3 columns]

In [27]:
long_data["value2"] = np.random.standard_normal(len(long_data))
long_data[:10]

date     item     value    value2
0 1959-01-01  realgdp  2710.349 -0.912937
1 1959-01-01     infl     0.000  0.227473
2 1959-01-01    unemp     5.800  1.191065
3 1959-04-01  realgdp  2778.801  0.345800
4 1959-04-01     infl     2.340 -2.020311
5 1959-04-01    unemp     5.100  1.015413
6 1959-07-01  realgdp  2775.488 -0.319257
7 1959-07-01     infl     2.740 -0.450373
8 1959-07-01    unemp     5.300 -0.444758
9 1959-10-01  realgdp  2785.204  0.105598

In [29]:
pivoted = long_data.pivot(index="date", columns="item") #hierarchical indexing
pivoted

value                     value2                    
item        infl    realgdp unemp      infl   realgdp     unemp
date                                                           
1959-01-01  0.00   2710.349   5.8  0.227473 -0.912937  1.191065
1959-04-01  2.34   2778.801   5.1 -2.020311  0.345800  1.015413
1959-07-01  2.74   2775.488   5.3 -0.450373 -0.319257 -0.444758
1959-10-01  0.27   2785.204   5.6  1.816850  0.105598  0.182082
1960-01-01  2.31   2847.699   5.2 -0.588279 -0.665956 -0.772348
...          ...        ...   ...       ...       ...       ...
2008-07-01 -3.16  13324.600   6.0 -0.335136 -1.213360  0.273298
2008-10-01 -8.79  13141.920   6.9  0.045058 -2.354614  0.626227
2009-01-01  0.94  12925.410   8.1 -0.297060 -0.357179  1.078032
2009-04-01  3.37  12901.504   9.2 -0.813674 -0.154461 -1.393434
2009-07-01  3.56  12990.341   9.6  0.746398  0.299843 -0.681089

[203 rows x 6 columns]

In [31]:
# unstack is equivalent to pivot 
unstacked = long_data.set_index(["date", "item"]).unstack(level="item")
unstacked

value                     value2                    
item        infl    realgdp unemp      infl   realgdp     unemp
date                                                           
1959-01-01  0.00   2710.349   5.8  0.227473 -0.912937  1.191065
1959-04-01  2.34   2778.801   5.1 -2.020311  0.345800  1.015413
1959-07-01  2.74   2775.488   5.3 -0.450373 -0.319257 -0.444758
1959-10-01  0.27   2785.204   5.6  1.816850  0.105598  0.182082
1960-01-01  2.31   2847.699   5.2 -0.588279 -0.665956 -0.772348
...          ...        ...   ...       ...       ...       ...
2008-07-01 -3.16  13324.600   6.0 -0.335136 -1.213360  0.273298
2008-10-01 -8.79  13141.920   6.9  0.045058 -2.354614  0.626227
2009-01-01  0.94  12925.410   8.1 -0.297060 -0.357179  1.078032
2009-04-01  3.37  12901.504   9.2 -0.813674 -0.154461 -1.393434
2009-07-01  3.56  12990.341   9.6  0.746398  0.299843 -0.681089

[203 rows x 6 columns]

In [33]:
df = pd.DataFrame({"key": ["foo", "bar", "baz"],
                   "A": [1, 2, 3],
                   "B": [4, 5, 6],
                   "C": [7, 8, 9]})
df

key  A  B  C
0  foo  1  4  7
1  bar  2  5  8
2  baz  3  6  9

In [34]:
melted = pd.melt(df, id_vars="key")
melted

key variable  value
0  foo        A      1
1  bar        A      2
2  baz        A      3
3  foo        B      4
4  bar        B      5
5  baz        B      6
6  foo        C      7
7  bar        C      8
8  baz        C      9

In [37]:
# `pd.melt` and `pivot` are inverses of each other
reshaped = melted.pivot(index="key", columns="variable", values="value")
reshaped

variable  A  B  C
key              
bar       2  5  8
baz       3  6  9
foo       1  4  7

In [38]:
pd.melt(df, id_vars="key", value_vars=["A", "B"]) # specify which columns to use

key variable  value
0  foo        A      1
1  bar        A      2
2  baz        A      3
3  foo        B      4
4  bar        B      5
5  baz        B      6